### PR TITLE
[SPARK-33480][SQL] Support char/varchar type

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR adds the char/varchar type which is kind of a variant of string type:
1. Char type is fixed-length string. When comparing char type values, we need to pad the shorter one to the longer length.
2. Varchar type is string with a length limitation.

To implement the char/varchar semantic, this PR:
1. Do string length check when writing to char/varchar type columns.
2. Do string padding when reading char type columns. We don't do it at the writing side to save storage space.
3. Do string padding when comparing char type column with string literal or another char type column. (string literal is fixed length so should be treated as char type as well)

To simplify the implementation, this PR doesn't propagate char/varchar type info through functions/operators(e.g. `substring`). That said, a column can only be char/varchar type if it's a table column, not a derived column like `SELECT substring(col)`.

To be safe, this PR doesn't add char/varchar type to the query engine(expression input check, internal row framework, codegen framework, etc.). We will replace char/varchar type by string type with metadata (`Attribute.metadata` or `StructField.metadata`) that includes the original type string before it goes into the query engine. That said, the existing code will not see char/varchar type but only string type.

char/varchar type may come from several places:
1. v1 table from hive catalog.
2. v2 table from v2 catalog.
3. user-specified schema in `spark.read.schema` and `spark.readStream.schema`
4. `Column.cast`
5. schema string in places like `from_json`, pandas UDF, etc. These places use SQL parser which replaces char/varchar with string already, even before this PR.

This PR covers all the above cases, implements the length check and padding feature by looking at string type with special metadata.

### Why are the changes needed?

char and varchar are standard SQL types. varchar is widely used in other databases instead of string type.

### Does this PR introduce _any_ user-facing change?

For hive tables: now the table insertion fails if the value exceeds char/varchar length. Previously we truncate the value silently.

For other tables:
1. now char type is allowed.
2. now we have length check when inserting to varchar columns. Previously we write the value as it is.

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3093]: We remove the char/varchar metadata after length check expressions are added, so that we don't do it repeatedly and this rule is idempotent.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala:44]: same here, we remove the char/varchar metadata to make this rule idempotent.

**Comment by cloud-fan** [comment]
cc @dongjoon-hyun @viirya @maropu

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1003]: This comes from https://github.com/apache/spark/commit/339eec5f321fac07e01441a3fd83024f4b92a0d6#diff-45c9b065d76b237bcfecda83b8ee08c1ff6592d6f85acca09c0fa01472e056afR1009 but is actually useless. The test still pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:27]: Shall we add `SPARK-6412 (Add Char support in dataTypes)` into the PR title as a secondary JIRA?
- SPARK-6412 was resolved as `Later`.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:27]: The PR description says `To be safe, this PR doesn't add char/varchar type to the core type system.`, but this adds `CharType` into `org.apache.spark.sql.types`. Do you revise a little the PR description? It seems that you wanted to mention the SQL execution parts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala:151]: `with string type` -> `with string type with metadata`?

**Comment by dongjoon-hyun** [comment]
cc @wangyum and @MaxGekk , too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:103]: To avoid accidentally adding the length check exprs again, we cannot remove the metadata at the same time as adding the exprs?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:157]: nit: How about extract `errorMsg` as a private method?
```
  private def raiseError(expr: Expression, typeName: String, length: Int): Expression = {
    val errorMsg = Concat(Seq(
      Literal("input string '"),
      expr,
      Literal(s"' exceeds $typeName type length limitation: $length")))
    Cast(RaiseError(errorMsg), StringType)
  }

  private def stringLengthCheck(expr: Expression, dt: DataType): Expression = dt match {
    case CharType(length) =>
      val trimmed = StringTrimRight(expr)
      // Trailing spaces do not count in the length check. We don't need to retain the trailing
      // spaces, as we will pad char type columns/fields at read time.
      If(
        GreaterThan(Length(trimmed), Literal(length)),
        raiseError(expr, "char", length),
        trimmed)
...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:103]: nit: `strLenChecked` -> `exprWithStrLenCheck`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:481]: Could you inline this in L476?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:97]: We can not, as the parser that is used to parse the schema string replaces char/varchar with string. This is the existing behavior.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:103]: > To avoid accidentally adding the length check exprs again, we cannot remove the metadata at the same time as adding the exprs?

We can't access the table relation here, only table output attrs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:27]: It's already there: https://github.com/apache/spark/pull/30412/files#diff-aacc5ed42589c636615a3c09a44fa6a5248195242c9fbe0e996db17471cd35fdL70

I just moved it to a new file and remove the parent class.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:27]: And type system means a lot more, e.g. expression type check, InternalRow.get, etc. Maybe I should say "type system framework"?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/VarcharType.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/VarcharType.scala:26]: VarcharType was already there: https://github.com/apache/spark/pull/30412/files#diff-aacc5ed42589c636615a3c09a44fa6a5248195242c9fbe0e996db17471cd35fdL79

I think it's a good idea to check the negative value, but adding max length should be done separately.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3093]: No it doesn't. Metadata is fine as it's harmless. We only need to watch out for some specific rules that look at the char/varchar metadata, and make sure they are idempotent.

As a fact, the added cast and length check expression is wrapped by an `Alias` which retains char/varchar metadata. So the output attributes of `Project` above the v2 relation still have metadata. It's necessary as we need to rely on it later to do padding for char type column comparison.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:97]: But it's safer to check all the nodes, I'll update it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:103]: > We can't access the table relation here, only table output attrs.

Oh, I see.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:27]: PR description updated

**Review by MaxGekk — COMMENTED** [review]
@cloud-fan You should rebase on the master to see the errors https://github.com/apache/spark/pull/30412#issuecomment-730424852 locally. The check came from https://github.com/apache/spark/pull/30351

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:33]: How about checking an output schema, too, in these tests?
```
scala> sql("CREATE TABLE t(i STRING, c CHAR(5)) USING parquet PARTITIONED BY (c)")
scala> spark.table("t").printSchema
root
 |-- i: string (nullable = true)
 |-- c: string (nullable = true) <---- this check
```
btw, how do users check a char length after defining a table? In pg, users can check a char length via some commands, e.g., `\d`;
```
postgres=# create table t (c char(5));
CREATE TABLE
postgres=# \d t
                    Table "public.t"
 Column |     Type     | Collation | Nullable | Default 
--------+--------------+-----------+----------+---------
 c      | character(5) |           |          | 
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:99]: In the case below, could we use `AnalysisException` instead?
```
scala> sql("""SELECT from_json("{'a': 'aaa'}", "a char(3)")""").printSchema()
java.lang.IllegalStateException: [BUG] logical plan should not have output of char/varchar type: Project [from_json(StructField(a,CharType(3),true), {'a': 'aaa'}, Some(Asia/Tokyo)) AS from_json({'a': 'aaa'})#37]
+- OneRowRelation
```

**Review by maropu — COMMENTED** [review]
Do you have a plan to update the doc: https://github.com/apache/spark/blob/master/docs/sql-ref-datatypes.md ? It would be nice to describe something about char/varchar there, I think.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:99]: I choose `IllegalStateException` because this should not happen. Thanks for catching this corner case and I've fixed it in `DataType.fromDDL`.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:346]: map type is not comparable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:99]: removed.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:33]: We ensure that `LocalPlan.output` won't have char/varchar type and we have UT to verify it. It seems not necessary to check it again here.

For DDL command support, can we leave it to followup? This PR is already very big. I can create a ticket for it.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:33]: The followup looks fine.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:343]: Ah, I see. If so, why the function `typeWithWiderCharLength` has an entry for `Map`? https://github.com/apache/spark/pull/30412/files#diff-16753285e80505e04c445ea8ccee1dde7ae601ed7bae224e212d90d395f57928R225-R226
It seems the function is only used for `addPaddingInStringComparison`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:129]: A question, do we need to support default length if user not give a special value ?

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:343]: You are right, I should remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:129]: We don't. We can set a default length/max length in followups.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:97]: @maropu I changed it back as it's pretty risky to get output from arbitrary logical plans. An example of the error:
```
Caused by: sbt.ForkMain$ForkError: java.lang.AssertionError: assertion failed: Scalar subquery should have only one column
	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.sql.catalyst.expressions.ScalarSubquery.dataType(subquery.scala:229)
	at org.apache.spark.sql.catalyst.expressions.Alias.toAttribute(namedExpressions.scala:181)
	at org.apache.spark.sql.catalyst.plans.logical.Project.$anonfun$output$1(basicLogicalOperators.scala:61)
```

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@maropu @dongjoon-hyun any more comments?

File: docs/sql-ref-datatypes.md
Comment (viirya) [inline docs/sql-ref-datatypes.md:41]: Reading column of type `VarcharType(n)` -> Reading column of type `CharType(n)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:140]: char type padding -> char type checking?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/VarcharType.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/VarcharType.scala:27]: if -> of

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/CharType.scala:28]: if -> of

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1186]: So we can do `cast(CharType)`? It actually casts to StringType? But don't we loss length info?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1186]: If you do `col.cast("char(5)")` before this PR, Spark already silently treats it as cast to string type. I don't want to change this behavior here. We can make it better later, by failing explicitly and saying that cast to char type is not supported.

**Comment by maropu** [comment]
okay, it looks good to me. Btw, this new feature will land in the v3.1 release? There are still some remaining works (e.g., https://github.com/apache/spark/pull/30412#discussion_r528884694 and https://github.com/apache/spark/pull/30412#discussion_r529253616) though. cc: @HyukjinKwon because of a v3.1 release manager

**Comment by cloud-fan** [comment]
Not that, before this PR, people can already create varchar type column but it has no length check during write. People can already create char type column with hive table but it truncates input string during write. This PR doesn't make things worse.

**Comment by cloud-fan** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan and all!

**Pull request closed without merging**
